### ISSUE TITLE
color-pick-contrast() does not always return the right color

### ISSUE THREAD

**Issue by katzmo** [foundation/foundation-sites#12086]
## What should happen?

I'd expect `color-pick-contrast()` to always return the color with the most contrast if above the given threshold.

E.g. in the following example, I'd expect the result to be `#FE0026`. `#FE0026` has as color-contrast of 4 which is above the threshold of 3, while `#ffffff` on `#ffffff` has the worst possible contrast 1.

```scss
@debug color-pick-contrast(#fff, (#fff, #FE0026), 3);
// expected: #FE0026
// actual: #ffffff
```

## What happens instead?

Instead, it returns the first color: `#ffffff`.

This is due to the [if statement in the function](https://github.com/foundation/foundation-sites/blob/v6.6.3/scss/util/_color.scss#L71):

```scss
@if ($current-contrast - $contrast > $tolerance) 
```

In the above example `$current-contrast` is 4 and `$contrast` is 1, so `4 - 1 = 3` and `3 > 3 != true`. 

## Possible Solution

Maybe this modification of the @if condition would do?

```scss
@if ($current-contrast - $contrast > 0 && $current-contrast > $tolerance) 
```

## Your Environment

Foundation version(s) used: 6.6.3

**Comment by DanielRuf** [comment]
Hi @katzmo,

thanks for letting us know. Can you provide a PR which fixes this?

If possible, tests would be also great (which fail without the changes and which succeed with the changes / fixes). See https://github.com/foundation/foundation-sites/blob/develop/test/sass/_color.scss

**Comment by katzmo** [comment]
Hi @DanielRuf!

I could give it try! But I now think I just didn’t understand how `$tolerance` is supposed to work?

I’ve found [this comment](https://github.com/foundation/foundation-sites/issues/10787#issuecomment-349309005) by @SassNinja explaining

> The tolerance is used here to specify how much the color that is currently checked (e.g. white) has to be better than the previous match.

So if I take this to mean the function should return
- the first color where the contrast is `$tolerance` greater than the contrast of first color or the previous color where the contrast was `$tolerance` greater etc
- the first color if there is no other color with a contrast greater than that of the first color + `$tolerance`

it does what it should, although I can’t quite see when this would be useful?

**Comment by DanielRuf** [comment]
If things can be impoved a change and explanation in the comments with test casesvare very helpful.

**Comment by katzmo** [comment]
Agree, but without a proper understanding of the idea behind `$tolerance` I can’t decide what’s an improvement and what is just a breaking change… and I guess anything could be a breaking change for people relying on the current implementation.

**Comment by apriljunge** [comment]
@katzmo You mentioned the [explenation of SassNinja](https://github.com/foundation/foundation-sites/issues/10787#issuecomment-349309005). I also read it and I think I quite understood it.

I think you got confused with the naming _tolerance_. It's not your thresold! For your case you should use a `$tolerance` of 0 (as SassNinja mentioned). Then all should work fine.

@DanielRuf maybe the information to (normally) keep tolerance of 0 should be added in the docs, as it might be confusing?

**Comment by DanielRuf** [comment]
Good idea @apriljunge :+1:

**Comment by katzmo** [comment]
Thanks @apriljunge! Would you by any chance know of a use case for $tolerance > 0?

**Comment by apriljunge** [comment]
@katzmo I think this option is to priorize first defined colors in a map. This would possibly make more sense if you have more than two colors defined.

For example:

These two colors do have the almost the same lightness. I prefer having primary over secondary color, if the contrast-difference is 1 or lower (`$tolerance: 1`). With this config, primary is chosen, even if secondary would be _slightly_ better.

```scss
$colors: (
  #40BF40, // primary
  #4095BF // secondary
);

.selector {
  attr: color-pick-contrast(white, $colors, $tolerance: 0); // returns #4095BF (secondary)
  attr: color-pick-contrast(white, $colors, $tolerance: 1); // returns #40BF40 (primary)
}
```

**Comment by katzmo** [comment]
Thanks for the example, I can see how this might be useful! 

Although I think it only works well for 2 colors. With many colors, the same logic is applied to each intermediate result, making the outcome less predictable: say a list of colors has contrasts of [1,2,3,4,5,6], `$tolerance: 1` will pick the color with a contrast value of 5 although 3 would be the first sufficient contrast and 6 the highest.

**Comment by apriljunge** [comment]
Hm, quite a good point! Now it gets complicated.
If I understand your case right `$tolerance` wouldn't be the best option to pass.
Maybe an option for the "at least contrast" would be better?

Or deprecating the whole option as this could be too project depending for a framework like this?

**Comment by katzmo** [comment]
Personally I think a better documentation of `$tolerance` would help most. "Contrast tolerance" is not clear at all, especially since there are no usage examples anywhere in Foundation. I admit it’s hard to describe the rather complex behavior in just a few words. Maybe something along the lines of "Contrast difference a color needs to surpass to beat a previous color in the compare list"?

As for your suggested option, having a function that returns the first color of a list that has a contrast higher than n seems very useful to me (in fact, I'm now using a similar custom function in my project). But I don’t think that’s compatible with `color-pick-contrast()` which I'd expect to return the color with the highest contrast, not the first in the list matching some criteria. I'm also conscious that adding everything that's nice to have to a framework like Foundation would eventually make it unusable complex.
